compiler/optimizations/optimizeOnClauses.cpp

```diff
@@ -411,30 +446,77 @@ optimizeOnClauses(void) {
   if (fNoOptimizeOnClauses)
     return;
 
+  // If we're not using locks, the extern functions in
+  // atomics are local and safe for fast on.
+  if (0 != strcmp(CHPL_ATOMICS, "locks")) {
+    forv_Vec(ModuleSymbol, module, gModuleSymbols) {
+      if( module->hasFlag(FLAG_ATOMIC_MODULE) ) {
+        Vec<FnSymbol*> moduleFunctions = module->getTopLevelFunctions(true);
+        forv_Vec(FnSymbol, fn, moduleFunctions) {
+          if( fn->hasFlag(FLAG_EXTERN) ) {
+            fn->addFlag(FLAG_FAST_ON_SAFE_EXTERN);
+            fn->addFlag(FLAG_LOCAL_FN);
+          }
+        }
+      }
+    }
+  }
```

Comment:
@mppf you're right that a sys call probably won't be involved. It's been a while since I looked through the libatomic source code, but I'm pretty sure it just does a CAS loop when hardware support isn't availble.

I don't have a real sense of what's legal in the AM handler though, so I'll leave that to you and Greg. Just wanted to note that intrinsics might not be lock-free and I don't think that note from the gcc docs implies anything about the processor support being lock-free. In particular gcc 4.8 (with libatomic) supports 64 bit atomics on 32 bit platforms, which is obviously not implemented using hardware, and is not lock-free (and gcc doesn't produce any warnings.)